### PR TITLE
Enable peeling for fully dynamic GEMMs and element-wise ops

### PR DESCRIPTION
**Pull request by dcaballe**
Running some performance experiments for now

### CODE REVIEW

**Comment by dcaballe** [comment]
Performance numbers:

|Micro-benchmark             | No Vectorization | Padding | Peeling |
| --------------------------- | ------------------- | --------- | -------- |
|  Dynamic GEMM              | 9.47 ms               |  0.88ms*  |  1.44ms |
| Dynamic Element-wise    |   7.30 ms             |  7.30ms**  |  2.10ms |

*Latency is lower with padding on GEMM because it vectorizes three dimensions, resulting in aggressive unrolling, and peeling is only vectorizing one (no unrolling). 
** Padding is not implemented for element-wise ops.

Let me know what you think.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -392,6 +430,46 @@ static void setAlwaysVectorizeSizes(linalg::LinalgOp op,
...
+          assert(reductionSizes[i] == 1 &&
+                 "This tile size should have been set to one");
+          reductionSizes[i] = origReductionSizes[i];
+          break;
+        }
+      }
+    }
+  }
+
+  return;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:470]: Let's make the level of nest lower. We can exit directly when it does not meet the requirements. E.g.,

```
if (!isFullyDynamicOp(op) || !isPeelingBeneficial(op)) return;
// Check isParallelDimVectorized...
if (isParallelDimVectorized) return;
// Check reduction cases..
```

**Review by hanhanW — APPROVED** [review]
Overall looks nice to me!

**Review by MaheshRavishankar — COMMENTED** [review]
Hard for me to follow the code (the logic here is getting pretty complicated :( , we might need to consolidate things in the future). What is the default mode for GEMMs with dynamic shapes or elementwise ops with dynamic shapes? Is it peeling or padding.

**Comment by dcaballe** [comment]
> Hard for me to follow the code (the logic here is getting pretty complicated :( , we might need to consolidate things in the future). 

I totally agree! I've been scattering comments about the same concerns in the last code reviews. I think we should follow a more principled approach and have a single source of truth for some of the codegen decisions/heuristics.  

> What is the default mode for GEMMs with dynamic shapes or elementwise ops with dynamic shapes? Is it peeling or padding.

GEMMs and element-wise ops that are fully dynamic (i.e., all dims are dynamic) are using peeling with this patch. Otherwise, padding is used. We modify the tile sizes so that we only vectorize the fastest-varying parallel dimension, if any, or, otherwise, the fastest varying reduction dimension. This is mostly to minimize the impact of peeling higher-order dimensions.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**